src/Optimizer.ts

```diff
@@ -0,0 +1,147 @@
+    this.reuseComponentsReport.sort(this.sortFunction);
+    const removeComponents = new RemoveComponents(this.componentProvider);
+    this.removeComponentsReport = removeComponents.getReport();
+    this.removeComponentsReport.sort(this.sortFunction);
+    const moveToComponents = new MoveToComponents(this.componentProvider);
+    this.moveToComponentsReport = moveToComponents.getReport();
+    this.moveToComponentsReport.sort(this.sortFunction);
+
+    return {
+      reuseComponents: this.reuseComponentsReport,
+      removeComponents: this.removeComponentsReport,
+      moveToComponents: this.moveToComponentsReport
+    };
+  }
+
+  /**
+   * this sort function can be used by .sort function to sort the {@link ReportElement} arrays.
+   *
+   *
+   */
+  private sortFunction = (a: ReportElement, b: ReportElement): number => { return (a.action.length - b.action.length || b.path.length - a.path.length);};
+
+  /**
+   * This function is used to get the optimized document after seeing the report.
+   *
+   * @param {Options} options - the options are a way to customize the final output.
+   * @returns {string } returns an stringified version of the YAML output.
+   *
+   */
+  getOptimizedDocument = (options: Options): string => {
+    const defaultOptions = {
+      rules: {
+        reuseComponents: true,
+        removeComponents: true,
+        moveToComponents: true
+      }
+    };
+    options = merge(defaultOptions, options);
+    this.outputObject = YAML.parse(this.YAMLorJSON);
+    if (options.rules?.reuseComponents) {
+      this.applyChanges(this.reuseComponentsReport);
+    }
+    if (options.rules?.moveToComponents) {
+      this.applyChanges(this.moveToComponentsReport);
+    }
+    if (options.rules?.removeComponents) {
+      this.applyChanges(this.removeComponentsReport);
+    }
+
+    return YAML.stringify(this.outputObject);
+  }
+
+  /**
+   * Sometimes removing and optimizing components leaves the parent empty and an empty object is of no use. this function
+   * checks if the parent is empty or not, if empty it will remove it.
+   *
+   * @param {string} childPath - the path of the child that we need to check its parent.
+   * @returns {void}
+   *
+   */
+  private removeEmptyParent = (childPath: string): void => {
+    const parentPath = childPath.substr(0, childPath.lastIndexOf('.'));
+    const parent = _.get(this.outputObject, parentPath);
+    if (Object.keys(parent).length === 0) {
+      _.unset(this.outputObject, parentPath);
+    }
+  }
+  /**
+   * this function will check if a component has parent or is a $ref to another component.
+   *
+   * @param {string} childPath - the path of child.
+   * @returns {void}
+   *
+   */
+  private hasParent = (childPath: string): boolean => {
+    const parentPath = childPath.substr(0, childPath.lastIndexOf('.'));
+    const parent = _.get(this.outputObject, parentPath);
+    return !(!parent || _.has(parent, '$ref'));
```

Comment:
`_.has` will return false even if `parent` is undefined. So you don't need to additionally check `!parent`